gk/main.c

```diff
@@ -478,6 +479,70 @@ get_block_idx(struct gk_config *gk_conf, unsigned int lcore_id)
 	return 0;
 }
 
+static bool
+is_flow_expired(struct flow_entry *fe, uint64_t request_timeout_cycles)
+{
+	uint64_t now = rte_rdtsc();
+
+	switch(fe->state) {
+	case GK_REQUEST:
+		if (fe->u.request.last_packet_seen_at > now) {
+			char err_msg[128];
+			int ret = snprintf(err_msg, sizeof(err_msg),
+				"gk: buggy condition at %s wrong timestamp",
+				__func__);
+			RTE_VERIFY(ret > 0 && ret < (int)sizeof(err_msg));
+			print_flow_err_msg(&fe->flow, err_msg);
+			return true;
+		}
+
+		return now - fe->u.request.last_packet_seen_at >
+			request_timeout_cycles;
+	case GK_GRANTED:
+		return now > fe->u.granted.cap_expire_at;
+	case GK_DECLINED:
+		return now > fe->u.declined.expire_at;
+	default:
+		return true;
+	}
+}
+
+static void
+gk_flow_tbl_bucket_scan(uint32_t *bidx, uint32_t *next,
+	uint64_t request_timeout_cycles, struct gk_instance *instance)
+{
+	int32_t index;
+	const struct ip_flow *key;
+	void *data;
+	time_t now = time(NULL);
+	RTE_VERIFY(now >= 0);
+
+	index = rte_hash_bucket_iterate(instance->ip_flow_hash_table,
+		(void *)&key, &data, bidx, next);
+	while (index >= 0) {
+		struct flow_entry *fe = &instance->ip_flow_entry_table[index];
+		if (is_flow_expired(fe, request_timeout_cycles)) {
+			int ret = rte_hash_del_key(
```

Comment:
The entire code block of this `if` statement is repeated in `drop_flow_entry_heuristically()`. Turn it into a function and avoid the duplicate code.